.tools/readmes/scanner.py

```diff
@@ -3,179 +3,145 @@
+                self.entities[svc.short] = svc.expanded.short
+        # config entities override
+        for entity, expanded in config.entities.items():
+            self.entities[entity] = expanded
+
+    def load_crosses(self):
+        self.doc_gen.process_metadata(
+            self.doc_gen.root / ".doc_gen" / "metadata" / "cross_metadata.yaml"
+        )
 
-    def _load_meta(self, file_name, field):
-        if field is not None:
-            return field
-        sdk_file_name = f"{self.meta_folder}/{file_name}"
-        with open(sdk_file_name, encoding="utf-8") as sdk_file:
-            meta = yaml.safe_load(sdk_file)
-        return meta
-
-    def _load_sdks(self):
-        self.sdk_meta = self._load_meta("sdks.yaml", self.sdk_meta)
-
-    def _load_services(self):
-        self.svc_meta = self._load_meta("services.yaml", self.svc_meta)
-
-    def _load_cross(self):
-        self.cross_meta = self._load_meta("cross_metadata.yaml", self.cross_meta)
+    def set_service(self, service):
+        self.svc_name = service
 
-    def _load_examples(self):
-        self.example_meta = self._load_meta(
-            f"{self.svc_name}_metadata.yaml", self.example_meta
-        )
+    def _build_examples(self):
+        self.examples = defaultdict(list)
+        for example in self.doc_gen.examples.values():
+            for lang_name, language in example.languages.items():
+                for sdk_version in language.versions:
+                    for svc_name in example.services:
+                        self.examples[
+                            f"{lang_name}:{sdk_version.sdk_version}:{svc_name}"
+                        ].append(example)
 
-    def _contains_language_version(self, example):
-        if self.lang_name in example["languages"]:
-            for version in example["languages"][self.lang_name]["versions"]:
-                if version["sdk_version"] == self.sdk_ver:
-                    return True
-        return False
+    def _example_key(self):
+        return f"{self.lang_name}:{self.sdk_ver}:{self.svc_name}"
 
-    def set_example(self, language, sdk_ver, service):
+    def set_example(self, language, sdk_ver):
         self.lang_name = language
         self.sdk_ver = sdk_ver
-        self.svc_name = service
-        self.example_meta = None
 
-    def sdk(self):
-        self._load_sdks()
-        return self.sdk_meta[self.lang_name]
+        self.hellos.clear()
+        self.actions.clear()
+        self.scenarios.clear()
+        self.customs.clear()
+        self.cross_scenarios.clear()
+        self.crosses.clear()
+
+        key = self._example_key()
+        examples = self.examples[key]
+        for example in examples:
+            if example.id.startswith("cross_"):
+                if example.category == config.categories["scenarios"]:
+                    self.cross_scenarios[example.id] = example
+                else:
+                    self.crosses[example.id] = example
```

Comment:
This might be safe, because likely this convention has been followed for naming, but the "true" categorization used by zexi is based off of the number of services (for Action vs. Cross) plus overrides.
Because this is just for README generation, it's probably okay if there are edge cases that don't quite line up.

Reply:
Since this is the same logic as WRITEME currently uses, if we want to change it, we can review after this PR.